### PR TITLE
feat(delegate-task): emit subagent_type enum when subagent list is known

### PR DESCRIPTION
**Pull request by WarGloom**
## Summary

- Builds the `task` tool argument `subagent_type` as a `zod.enum(...)` when the runtime can enumerate available subagents, instead of a free-form string.
- Falls back to plain string when no list is available, preserving current behavior.

## Changes

- New optional field on `DelegateTaskToolOptions`: `availableSubagentNames?: readonly string[]`.
- `createDelegateTask` builds its args schema dynamically:
  - non-empty list → `tool.schema.enum(names).optional().describe(...)`
  - otherwise → existing `tool.schema.string().optional().describe(...)`
- `tool-registry.ts` plumbs the discovery: union of `BUILTIN_SUBAGENT_TYPES` + agents from `loadUserAgents`, `loadProjectAgents`, `loadOpencodeGlobalAgents`, `loadOpencodeProjectAgents`, `readOpencodeConfigAgents`. Filters by `mode === "subagent" | "all" | undefined`. Removes entries listed in `pluginConfig.disabled_agents`. Deduplicated.

The description text, runtime branches, category resolution, and execution paths are unchanged.

## Motivation

Today `subagent_type` is `tool.schema.string().optional()`. The resulting JSON Schema only carries `{"type":"string"}` — there is no machine-readable list of valid agent names anywhere in the tool definition. Downstream consumers that do not parse the description text (LLM proxies, validators, harness adapters, observability tools) cannot discover the user's configured subagents.

JSON Schema's `enum` is the canonical place to declare a closed set of string values, and it is what every standard tooling pipeline already understands.

## Backward compatibility

- Public option is fully optional; no caller is required to change.
- When `availableSubagentNames` is omitted or empty, the schema is byte-identical to before.
- The `category`, `subagent_type`, and `task_id` runtime branches in the tool body are untouched.
- No change to category configs, fallback chains, or sync/background execution.

## Testing

```bash
bun run typecheck
bun test src/tools/delegate-task/task-schema.test.ts
```

3 new tests:

- No `availableSubagentNames` → `subagent_type` stays plain string.
- `availableSubagentNames: ["oracle", "librarian", "dev"]` → `subagent_type` is enum and contains those names.
- Empty array → falls back to plain string.

## Token impact

Each agent name adds ~3-5 input tokens to the tool definition. The tool definition lives in the cached prompt prefix, so the cost is paid once and amortizes to near-zero across a session.

## Why this matters

Without an enum on `subagent_type`, any harness sitting between OpenCode and the model has to parse natural-language description text to discover the agent list — fragile and version-dependent. With it, the canonical JSON Schema field carries the list, which is what every standard tooling pipeline already knows how to consume.

---
<a href="https://app.blacksmith.sh/code-yeongyu/codesmith/oh-my-openagent/pr/3719"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-in-codesmith-light.svg">View in Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>@codesmith</code> with what you need.</sup>

- [ ] Let Codesmith autofix CI failures and bot reviews

---
## Summary by cubic
Makes `task.subagent_type` a constrained enum when the runtime knows available subagents, improving validation and tooling. Falls back to a plain string when unknown; the registry now discovers, filters, and passes subagent names.

- **New Features**
  - Emit `zod.enum(...)` for `subagent_type` when `availableSubagentNames` is non-empty; otherwise use a string.
  - Discover names from built-ins plus `loadUserAgents`, `loadProjectAgents`, `loadOpencodeGlobalAgents`, `loadOpencodeProjectAgents`, `readOpencodeConfigAgents`, and `loadAgentDefinitions`. Respect `pluginConfig.claude_code?.agents`, filter by `mode`, exclude `pluginConfig.disabled_agents`, dedupe/sort, and pass via `availableSubagentNames` to `createDelegateTask`.
  - Add `availableSubagentNames?: readonly string[]` to `DelegateTaskToolOptions`; add tests for enum, empty, and missing cases.

<sup>Written for commit f142b13091e69239c69836465e6164dcac7d8343. Summary will update on new commits. <a href="https://cubic.dev/pr/code-yeongyu/oh-my-openagent/pull/3719?utm_source=github">Review in cubic</a></sup>

### CODE REVIEW

File: src/plugin/tool-registry.ts [hunk 1]
```diff
@@ -178,10 +186,33 @@ export function createToolRegistry(args: {
   )
   const lookAt = isMultimodalLookerEnabled ? factories.createLookAt(ctx) : null
 
+  const disabledAgents = new Set((pluginConfig.disabled_agents ?? []).map((a) => a.toLowerCase()))
+  const agentDicts = [
+    loadUserAgents(),
+    loadProjectAgents(ctx.directory),
+    loadOpencodeGlobalAgents(),
+    loadOpencodeProjectAgents(ctx.directory),
+    readOpencodeConfigAgents(ctx.directory),
```
[the remark this answers is not in the document]
Comment (WarGloom) [reply src/plugin/tool-registry.ts:207]: Partially addressed in `f142b130`:

- `agent_definitions` agents now flow into the enum via `loadAgentDefinitions(pluginConfig.agent_definitions, "definition-file")`.
- Plugin manifest agents (`pluginComponents.agents`) NOT yet included — `pluginComponents` is loaded in the agent-config-handler pipeline and isn't currently threaded down to `createToolRegistry` / `createTools`. Plumbing it would expand the change beyond the schema-emission scope of this PR. Will follow up in a separate change that lifts `pluginComponents` into the tool-registry args. Tracking comment in code so the next person sees the gap.

File: src/plugin/tool-registry.ts [hunk 2]
```diff
@@ -178,10 +186,33 @@ export function createToolRegistry(args: {
   )
   const lookAt = isMultimodalLookerEnabled ? factories.createLookAt(ctx) : null
 
+  const disabledAgents = new Set((pluginConfig.disabled_agents ?? []).map((a) => a.toLowerCase()))
+  const agentDicts = [
+    loadUserAgents(),
+    loadProjectAgents(ctx.directory),
```
[the remark this answers is not in the document]
Comment (WarGloom) [reply src/plugin/tool-registry.ts]: Addressed in `f142b130`. The discovery now reads `pluginConfig.claude_code?.agents ?? true` and gates `loadUserAgents()` / `loadProjectAgents()` on that flag — same logic `applyAgentConfig` uses (`src/plugin-handlers/agent-config-handler.ts:106-108`). When `claude_code.agents` is disabled, those agents are no longer advertised by the schema, matching runtime registration.

File: src/plugin/tool-registry.ts [hunk 3]
```diff
@@ -178,10 +186,33 @@ export function createToolRegistry(args: {
...
+      const mode = (cfg as { mode?: string } | null | undefined)?.mode
+      if (mode === "subagent" || mode === "all" || mode === undefined) {
+        discoveredNames.add(name)
+      }
+    }
+  }
+  const builtinNames = BUILTIN_SUBAGENT_TYPES.map((t) => t.name)
+  const availableSubagentNames = Array.from(
+    new Set([...builtinNames, ...discoveredNames]),
+  ).filter((name) => !disabledAgents.has(name.toLowerCase()))
```
[the remark this answers is not in the document]
Comment (WarGloom) [reply src/plugin/tool-registry.ts]: Addressed in `f142b130`. Final list is sorted with `localeCompare` for deterministic output across runs.

**Comment by WarGloom** [comment]
## Review status

### In-scope (this PR's diff) — addressed in `f142b130`

- [x] **P1** — Include plugin and definition agents in subagent enum: `agent_definitions` agents now included via `loadAgentDefinitions(pluginConfig.agent_definitions, "definition-file")`. Plugin manifest agents (`pluginComponents.agents`) require threading `pluginComponents` through `createTools` → `createToolRegistry`; deferred to a follow-up PR to keep this change focused on schema emission.
- [x] **P2** — Respect `claude_code.agents` flag: `loadUserAgents()` / `loadProjectAgents()` are now gated on `pluginConfig.claude_code?.agents ?? true`, matching `applyAgentConfig`.
- [x] **P2** — Sort enum: final list sorted with `localeCompare` for deterministic output.

Also rebased to a single clean commit on `dev`.

### Out of scope (not in this PR's diff)

The following comments are flagged on file paths that aren't part of this PR's diff (they were attached during the initial review of the pre-rebase commit, which carried unrelated changes from the original branch):

- `src/hooks/compaction-context-injector/recovery.ts`
- `src/cli/run/agent-profile-colors.ts`
- `src/config/schema/experimental.ts`
- `src/shared/model-capabilities/supplemental-entries.ts`
- `src/shared/serena-availability.ts`

Those files are no longer touched by this PR after the rebase. Happy to file separate PRs for any of them if they're real bugs on `dev`.

**Comment by acamq** [comment]
Hi @WarGloom, thanks for your contribution!

This PR is being closed because the [CLA check](https://github.com/code-yeongyu/oh-my-openagent/actions/workflows/cla.yml) has been failing and the PR has been inactive for **7+ days**.

The CLA is a requirement for all contributions to this project. To resolve this:

1. Sign the CLA by following the instructions in the CLA check failure comment on your PR.
2. Reopen this PR (or open a new one targeting the latest `dev` branch) once signed.

If you believe this was closed in error, feel free to comment and we'll take another look.

Thanks again for your time and contribution!
